### PR TITLE
c2c: add SetupSpanConfigsStream(tenantName) to stream client

### PR DESCRIPTION
**Pull request by msbutler**
This patch adds a new call to the stream client, which uses the new
crdb_internal.setup_span_configs_stream() builtin, to create a replication
stream specification to stream span config updates relevant to the replicating
tenant. This new client call essentially combines the client.Create() and
client.Plan() into one call, as it returns a streamID and topology to the user.
Note this new call differs from client.Create() as it neither creates a job nor
lays a protected timestamp on the source cluster. In other words, the user
should treat this span configuration stream as ephemeral.

A future PR will teach client.Subscribe to stream these updates and ingest them
into the destination.

Informs https://github.com/cockroachdb/cockroach/issues/106823

Release note: None

### CODE REVIEW

**Comment by msbutler** [comment]
If you're curious, these two PRs are next in the queue:
- https://github.com/cockroachdb/cockroach/pull/107400
- https://github.com/cockroachdb/cockroach/pull/107557

File: pkg/ccl/streamingccl/streamproducer/stream_lifetime.go [hunk 1]
```diff
@@ -307,3 +326,37 @@ func completeReplicationStream(
...
+		return nil, err
+	}
+	codec := keys.MakeSQLCodec(roachpb.SystemTenantID)
+	spanConfigKey := codec.TablePrefix(uint32(spanConfigID))
+
+	// TODO(msbutler): crop this span to the keyspan within the span config
+	// table relevant to this specific tenant once I teach the client.Subscribe()
+	// to stream span configs, which will make testing easier.
+	span := roachpb.Span{Key: spanConfigKey, EndKey: spanConfigKey.PrefixEnd()}
+	return buildReplicationStreamSpec(ctx, eval, tenantID, true, roachpb.Spans{span})
```
Comment (stevendanna) [inline pkg/ccl/streamingccl/streamproducer/stream_lifetime.go:361]: A couple of questions here:

1) In a recent conversation, I believe you and Arul pointed out to me that the span configuration table assumes some invariants.  Are we sure we are going to be able to deal with those invariants will be OK if we divide processing of the table up across. multiple nodes?  

2) Aside from the above, I wonder if we could get away without this?  Even if the span config table is split across multiple ranges, we could still stream it from a single source node, that source node would just have a couple remote rangefeeds potentially.  But, that doesn't seem _that_ bad to me.

File: pkg/ccl/streamingccl/streamproducer/replication_stream_test.go [hunk 2]
```diff
@@ -229,6 +230,42 @@ func testStreamReplicationStatus(
...
 }
 
+func TestSpanConfigReplicationStreamSetup(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	serverArgs := base.TestServerArgs{
+		// This test fails when run from within a test tenant. This is likely
+		// due to the lack of support for tenant streaming, but more
+		// investigation is required. Tracked with #76378.
```
Comment (stevendanna) [inline pkg/ccl/streamingccl/streamproducer/replication_stream_test.go:240]: Let's not add a TODOTestTenantDisabled here.  Let's use TestIsSpecificToStorageLayerAndNeedsASystemTenant or TestControlsTenantsExplicitly. Both seem true in this case.

File: pkg/ccl/streamingccl/streamproducer/replication_stream_test.go [hunk 3]
```diff
@@ -229,6 +230,42 @@ func testStreamReplicationStatus(
...
+		Knobs: base.TestingKnobs{
+			JobsTestingKnobs: jobs.NewTestingKnobsWithShortIntervals(),
+		},
+	}
+
+	h, cleanup := replicationtestutils.NewReplicationHelper(t, serverArgs)
+	defer cleanup()
+	testTenantName := roachpb.TenantName("test-tenant")
+	_, cleanupTenant := h.CreateTenant(t, serverutils.TestTenantID(), testTenantName)
+	defer cleanupTenant()
```
Comment (stevendanna) [inline pkg/ccl/streamingccl/streamproducer/replication_stream_test.go:251]: We don't really need to even start the tenant sql server for this this. I think we could just `CREATE TENANT $1;`

File: pkg/ccl/streamingccl/streamproducer/stream_lifetime.go [hunk 4]
```diff
@@ -307,3 +326,37 @@ func completeReplicationStream(
...
+	tenantRecord, err := sql.GetTenantRecordByName(ctx, eval.Settings, txn, tenantName)
+	if err != nil {
+		return nil, err
+	}
+
+	tenantID := roachpb.MustMakeTenantID(tenantRecord.ID)
+
+	var spanConfigID descpb.ID
+	execConfig := eval.Planner.ExecutorConfig().(*sql.ExecutorConfig)
+	if err := sql.DescsTxn(ctx, execConfig, func(ctx context.Context, txn isql.Txn, col *descs.Collection) error {
```
Comment (stevendanna) [inline pkg/ccl/streamingccl/streamproducer/stream_lifetime.go:343]: Does this run in the transaction we are already a part of? Should it?

File: pkg/ccl/streamingccl/streamproducer/stream_lifetime.go [hunk 5]
```diff
@@ -226,30 +231,44 @@ func getReplicationStreamSpec(
...
+	if forSpanConfigs {
+		spanConfigsStreamID = streampb.StreamID(builtins.GenerateUniqueInt(builtins.ProcessUniqueID(eval.NodeID.SQLInstanceID())))
+	}
+
 	res := &streampb.ReplicationStreamSpec{
-		Partitions:     make([]streampb.ReplicationStreamSpec_Partition, 0, len(spanPartitions)),
-		SourceTenantID: details.TenantID,
+		Partitions:         make([]streampb.ReplicationStreamSpec_Partition, 0, len(spanPartitions)),
+		SourceTenantID:     tenantID,
+		SpanConfigStreamID: spanConfigsStreamID,
```
Comment (stevendanna) [inline pkg/ccl/streamingccl/streamproducer/stream_lifetime.go:269]: What is this streamID going to be used for now that we've decided we don't need a producer-side job?

File: pkg/ccl/streamingccl/streamproducer/stream_lifetime.go (hunk 1 above)
Comment (msbutler) [reply pkg/ccl/streamingccl/streamproducer/stream_lifetime.go:361]: Good questions!! I needed to dig a bit more here.

This [comment](https://github.com/cockroachdb/cockroach/blob/00ec8d330a18d77e3aaa2e6d02e79f2d03a01404/pkg/spanconfig/spanconfig.go#L45) describes those invariants and this [little beast](https://github.com/cockroachdb/cockroach/blob/7a57e7ea1e4b188c76407edf4d318099a9f2bcbf/pkg/spanconfig/spanconfigstore/span_store.go#L284) of a function enforces them on all _proposed_ span config updates. For C2C to obey these invariants, I _believe_ we merely need to stream the point/range tombstones that the `kv.Accessor` writes to the source's span_configuration table, and not worry about manually splitting span config records based on what's already been committed.

To that end, C2C must stream all updates from the span_configuration table in the right order, even  if the table is split across ranges. (I should check with Arul if a given tenant's span config updates can get split across ranges, though I imagine this is allowed). More explicitly, if we write a tombstone on a tenant's span config update at t1, we must replicate that tombstone before replicating any updates later than t1, regardless of any simultaneous splits. 

A question for a rangefeed expert (you): Do we get this ordering guarantee for free, with the "rangefeed retry on splits" behavior, if one src node is running the show (i.e. your proposal in question 2)? Or would I need to add extra hwm timestamp management on the destination side?

File: pkg/ccl/streamingccl/streamproducer/stream_lifetime.go (hunk 4 above)
Comment (msbutler) [reply pkg/ccl/streamingccl/streamproducer/stream_lifetime.go:341]: > Does this run in the transaction we are already a part of?

This sql.DescsTxn() creates creates a new txn via the executorConfig, which I believe is independent of the `p.InternalSQLTxn()` that we [initialize ](https://github.com/cockroachdb/cockroach/blob/574a09e39025ae5021d9d224e061168fd01aeda3/pkg/sql/planner.go#L910)the replication manager with. 


> Should it?

I'm not totally sure. If the replication manager is _supposed_ to read from kv only via the `replicationManager.Txn`, part of the [replication manager state](https://github.com/cockroachdb/cockroach/blob/574a09e39025ae5021d9d224e061168fd01aeda3/pkg/ccl/streamingccl/streamproducer/replication_manager.go#L30), then we're being naughty:
1. https://github.com/cockroachdb/cockroach/blob/574a09e39025ae5021d9d224e061168fd01aeda3/pkg/ccl/streamingccl/streamproducer/stream_lifetime.go#L187
2. https://github.com/cockroachdb/cockroach/blob/574a09e39025ae5021d9d224e061168fd01aeda3/pkg/ccl/streamingccl/streamproducer/stream_lifetime.go#L222

My hunch is that we probably ought to use the same txn for the examples I list and in this PR, as its cleanest to read the db from one consistent timestamp. Do you agree? 

(FWIW, I was questioning why we even had this seperate stateful txn object, and [here's](https://github.com/cockroachdb/cockroach/pull/90964) why. )

**Comment by msbutler** [comment]
@stevendanna RFAL. Adding do-not-merge label to prevent myself from borsing this PR without squashing the latest commit.

File: pkg/ccl/streamingccl/streamclient/client.go [hunk 6]
```diff
@@ -51,6 +51,12 @@ type Client interface {
 	// can be used to interact with this stream in the future.
 	Create(ctx context.Context, tenant roachpb.TenantName) (streampb.ReplicationProducerSpec, error)
 
+	// StartAndPlanEphemeralSpanConfigsStream creates a stream for the span configs
+	// that apply to the passed in tenant, and returns the subscriptions the
+	// client can subscribe to. No protected timestamp or job is persisted to the
+	// source cluster.
+	StartAndPlanEphemeralSpanConfigsStream(ctx context.Context, tenant roachpb.TenantName) (streampb.StreamID, Topology, error)
```
Comment (stevendanna) [inline pkg/ccl/streamingccl/streamclient/client.go:58]: [nit] We don't need to make a final decision now, but I think that the naming here is a bit odd now that we aren't actually making any sort of producer-side job. Namely, this function doesn't _start_ anything.  And, in other parts of the code and comments we seem to want this call this `Create...` or `Setup...`.  Personally, I would go with something like `PlanSpanConfigsStream`.

**Review by stevendanna — APPROVED** [review]
Thanks for pushing this forward.  I've left a good number of comments but am still going to mark approve.  I think many of my comments are things we can consider as we continue to develop this over the week.

**Comment by msbutler** [comment]
TFTR!

bors r=stevendanna

**Pull request merged**